client/my-sites/plans-features-main/components/plan-upsell-modal/hooks/use-progress-blocking-model.tsx

```diff
@@ -0,0 +1,50 @@
+import { isFreePlan } from '@automattic/calypso-products';
+import useIsCustomDomainAllowedOnFreePlan from 'calypso/my-sites/plans-features-main/hooks/use-is-custom-domain-allowed-on-free-plan';
+import {
+	PAID_DOMAIN_FREE_PLAN_SELECTED_MODAL,
+	FREE_DOMAIN_FREE_PLAN_SELECTED_MODAL,
+	MODAL_LOADER,
+	PAID_DOMAIN_PAID_PLAN_REQUIRED,
+} from '..';
+type Props = {
+	paidDomainName?: string | null;
+	flowName?: string;
+};
+
+type ModalType =
+	| typeof PAID_DOMAIN_FREE_PLAN_SELECTED_MODAL
+	| typeof FREE_DOMAIN_FREE_PLAN_SELECTED_MODAL
+	| typeof PAID_DOMAIN_PAID_PLAN_REQUIRED
+	| typeof MODAL_LOADER;
+type ModalStateDetails = {
+	resolveDisplayedModal: ( currentSelectedPlan: string ) => ModalType | null;
+};
+
+export function useProgressBlockingModal( { paidDomainName, flowName }: Props ): ModalStateDetails {
+	const isCustomDomainAllowedOnFreePlan = useIsCustomDomainAllowedOnFreePlan(
+		flowName,
+		!! paidDomainName
+	);
+	const resolveDisplayedModal = ( currentSelectedPlan: string ): ModalType | null => {
+		if ( isFreePlan( currentSelectedPlan ) ) {
+			if ( ! paidDomainName ) {
+				return FREE_DOMAIN_FREE_PLAN_SELECTED_MODAL;
+			}
+
+			if ( isCustomDomainAllowedOnFreePlan ) {
+				return PAID_DOMAIN_FREE_PLAN_SELECTED_MODAL;
+			}
+
+			/**
+			 * Either this or the above modal to be removed
+			 * after experiment 21394-explat-experiment is over
+			 */
+			return PAID_DOMAIN_PAID_PLAN_REQUIRED;
+		}
+		return null;
+	};
+
+	return {
+		resolveDisplayedModal,
+	};
```

Comment:
What is the point of returning an object here and not the callback directly?

The names also don't match and look confusing. `useProgressBlockingModal` basically just returns the `resolveDisplayedModal` callback just inside an object. tbh neither of these is super clear to me, but at least stick with one maybe. Just returning the callback should eliminate one of them. It's almost like you are trying to force from here however the consumer end calls the local variable 😕